git-index/src/access.rs

```diff
@@ -1,6 +1,83 @@
+use crate::{
+    entry::{self, Flags, Mode, Stat, Time},
+    extension, Entry, PathStorage, State, Version,
+};
 use bstr::{BStr, ByteSlice};
+use git_object::{tree::EntryMode, TreeRefIter};
+use git_traverse::tree::breadthfirst;
 
-use crate::{entry, extension, Entry, PathStorage, State, Version};
+/// initialization
+impl State {
+    /// Takes in an oid of a tree object and creates and returns a [`State`][git_index::State] from its children.
+    pub fn from_tree<Find>(tree: &git_hash::oid, mut find: Find) -> Result<Self, breadthfirst::Error>
+    where
+        Find: for<'a> FnMut(&git_hash::oid, &'a mut Vec<u8>) -> Option<TreeRefIter<'a>>,
+    {
+        let mut buf = Vec::new();
+        let root = find(tree, &mut buf).expect("couldn't find a tree for given oid");
+        let state = breadthfirst::State::default();
+        let mut delegate = git_traverse::tree::Recorder::default();
+        breadthfirst(root, state, &mut find, &mut delegate)?;
+
+        // TODO: estimate size
+        let mut path_backing: PathStorage = Vec::new();
+
+        // TODO: insert entries directly into the correct position
+        // TODO: double check this way of sorting is actually the right one
```

Comment:
> Index entries are sorted in ascending order on the name field,
interpreted as a string of unsigned bytes (i.e. memcmp() order, no
localization, no special casing of directory separator '/'). Entries
with the same name are sorted by their stage field.

I believe this is how `BStr` sorts, so this should already be correct. However, I don't know if the intent is necessarily for `BStr` explicitly intends to always sort this way, so maybe it's worth either clarifying that upstream or otherwise explicitly dereferencing down to the byte slice and comparing that instead.